addons/base_address_structured/models/base_address_structured.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import re
+from odoo import api, fields, models
+
+address_pattern = re.compile('^(.*?),? ([0-9]+[a-zA-Z]*)([ -]+([0-9]+[a-zA-Z]*))?[ ]*$')
+
+
+class ResCountryStateCity(models.Model):
+    _name = 'res.country.state.city'
+    _order = 'name'
+
+    name = fields.Char('City Name', required=True)
+    state_id = fields.Many2one('res.country.state', 'State')
+    country_id = fields.Many2one('res.country', 'Country', related='state_id.country_id', store=True, readonly=True)
+    code = fields.Char('Code')
+
+
+class PartnerAddress(models.Model):
+    _inherit = ['res.partner']
+    _name = 'res.partner'
+
+    street_raw = fields.Char('Street')
+    street_number1 = fields.Char('Street Number')
+    street_number2 = fields.Char('Door Number')
```

Comment:
One possible foundation to handle addresses around the world: https://github.com/openvenues/pypostal